### ISSUE TITLE
ExtendedParallelCucumber.class appears to execute sequentially not in parallel

### ISSUE THREAD

**Issue by MarkWilkinsonWB** [mkolisnyk/cucumber-reports#80]
I'm trying out the Parallel test running feature but all my scenarios seem to be running sequentially. I must be doing something wrong. Here's my Class and configuration. I pass in features and tags via -Dcucumber.options="....". I've tried running several Scenarios in one .feature file and also tried running several scenarios across a set of .feature files. The system appears to spawn only one single thread each time and the scenarios are run one at a time.

Using v1.0.5 of the library.

`package com.whitbread.automation.common;

import com.github.mkolisnyk.cucumber.runner.ExtendedCucumberOptions;
import com.github.mkolisnyk.cucumber.runner.ExtendedParallelCucumber;
import cucumber.api.CucumberOptions;
import org.junit.runner.RunWith;

@RunWith(ExtendedParallelCucumber.class)
@ExtendedCucumberOptions(
        retryCount = 1
        ,threadsCount = 4
)
@CucumberOptions(
        features="override-this-on-command-line"
        ,glue={"com.whitbread.automation.common"}
        ,tags={"@override-this-on-command-line"}
        ,plugin={"json:./output.json"}
)

public class GenericExtendedParallelRunner {
}
`

**Comment by mkolisnyk** [comment]
By design the runner splits execution based on features. It means that entire feature is performed in separate thread. Also, currently I don't know if it works properly if you override some options from command line.

Another thing is that each thread is spawned with small delay (usually a few seconds), so if your tests are short or even empty they may stop before the engine recognises that it's time to add another process

**Comment by MarkWilkinsonWB** [comment]
Thanks for the info. Will do some further testing tomorrow using scenarios that share the same tag, but are in different .feature scripts. I'll let you know the results. If it's successful then I'll probably refactor my scripts into more features to get greater parallelism.

Great library by the way. I have still yet to investigate all the reporting possibilities as we have to have some way of showing test results in Confluence..

**Comment by mkolisnyk** [comment]
Sure, let me know if you find something. 

Regarding reporting possibilities, you can use [documentation site](http://mkolisnyk.github.io/cucumber-reports/) for this library. Of course, a lot of possibilities aren't documented yet but the work is in progress.

And if you need some more features feel free to create new issue here in GitHub.

**Comment by MarkWilkinsonWB** [comment]
Just doing some further testing. It seems that passing in the feature directory and the tags via the command line triggers different behaviour from hard-coding in the annotation on the class. When passing in via -Dcucumber.options the scenarios are run correctly, but sequentially. When I hard-code them in the annotation, and use 2 separate tags in 2 separate feature files, then the first scenario works and the second always seems to fail, leaving an instance of the browser open as if @After isn't being triggered which should close down the Selenium session.  I tried both the standard selenium-server-standalone (FIrefox) and Chrome. In addition, hard-coding seems to produce addition stdout output i.e.

None of the features at [C:\Users\Mark\workspace\test-automation-hoh\features\availability-checker.feature] matched the filters: [@hoh-bricklane,@hoh-cookiedismiss]

Whereas, passing the features path and tags in via -Dcucumber.options seems not to produce these messages.

I can't give you the details to repro this exactly as the software I'm creating is for a private Company and is subject to copyright but I'd try different combinations of tags to run embedded in the @CucumberOptions annotation.. i.e. one tag that is on scenarios in more than one .feature file and then 2 or more tags, each of which represents a scenario in 2 or more .feature files.

**Comment by mkolisnyk** [comment]
Let's take a look at your items one by one:

1) I'm definitely uncertain how the runner works when Cucumber options are passed via command line as I didn't test it. I suspect that the extended cucumber runner which is used for parallel run doesn't transfer all things properly. As the result, the test run is performed as if you didn't pass any options. That's my guess. If it is so, let's take it as the defect and I'll work on fixing it. So, for now you can define the set of tests explicitly in the test class to make things working.

2) **None of the features at [C:\Users\Mark\workspace\test-automation-hoh\features\availability-checker.feature] matched the filters: [@hoh-bricklane,@hoh-cookiedismiss]** - this happens when you specify some features which don't match the tag filter. It's OK.

3) With these parallel runs I spotted that threads after the first ones are definitely failing sometimes. This can be related to the collision when they try to acquire selenium instance simultaneously. But after that everything works OK. I tried multiple options one engine side but at the moment there's no success in it. What I did for my company project is to update the hook which initialises Selenium instance. The code approximately looks like this:

```java
    @Before
    public void setUp(Scenario scenario) throws Exception {
        Exception error = null;
            for (int i = 0; i < 5; i++) {
                try {
                    start();
                    return;
                } catch (Exception e) {
                    error = e;
                    stop();
                    Thread.sleep(5 * DateTimeConstants.MILLIS_PER_SECOND);
                }
            }
            if (error != null) {
                throw error;
            }
    }
``` 
Here the **start()** and **stop()** methods simply start and stop Selenium instance respectively.
This way we have several tries to initialise Selenium instance properly (with session close in case of error). In such combination it works. I know it's not the best solution for now but the problem is related to infrastructure rather than runner itself. Let's take it as another defect.

**Comment by oscar-mullin** [comment]
I would suggest looking at Courgetter runner as you can split execution based on scenarios in addition to features. Maybe you can get some ideas from there. The reporting is worse than Extended Runner though.
